### ISSUE TITLE
TypeError: Cannot read property 'then' of undefined

### ISSUE THREAD

**Issue by mschmidt42** [spruce/ep_small_list#6]
Hi,

I get this error in version 0.3.0. 0.2.0 works:

* Ubuntu 18.04
* Node v9.11.2
* Etherpad version 1.7.5

[2019-07-03 16:01:15.661] [ERROR] console - TypeError: Cannot read property 'then' of undefined
    at /opt/etherpad/node_modules/ep_small_list/index.js:8:33
    at Layer.handle [as handle_request] (/opt/etherpad/src/node_modules/express/lib/router/layer.js:95:5)
    at next (/opt/etherpad/src/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/opt/etherpad/src/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/opt/etherpad/src/node_modules/express/lib/router/layer.js:95:5)
    at /opt/etherpad/src/node_modules/express/lib/router/index.js:281:22
    at Function.process_params (/opt/etherpad/src/node_modules/express/lib/router/index.js:335:12)
    at next (/opt/etherpad/src/node_modules/express/lib/router/index.js:275:10)
    at /opt/etherpad/src/node/hooks/express/webaccess.js:85:20
    at authorize (/opt/etherpad/src/node/hooks/express/webaccess.js:24:51)

Markus

**Comment by spruce** [comment]
Sorry for the delay. What version of etherpad-light are you using?

**Comment by mschmidt42** [comment]
As described above it is 1.7.5:

`nodejs[30259]: [2019-07-29 10:21:19.854] [INFO] console - Your Etherpad version is 1.7.5`

**Comment by spruce** [comment]
I still can't reproduce the issue. The only thing that was changed in this version jump is that we use promises and they are supported by your node version.

**Comment by mschmidt42** [comment]
I completely reinstalled etherpad-lite, started ehterpad-lite, set an admin password and installed ep_small-list via the admin interface:

`# git clone --branch master https://github.com/ether/etherpad-lite.git
Cloning into 'etherpad-lite'...
remote: Enumerating objects: 81, done.
remote: Counting objects: 100% (81/81), done.
remote: Compressing objects: 100% (74/74), done.
remote: Total 32849 (delta 39), reused 31 (delta 7), pack-reused 32768
Receiving objects: 100% (32849/32849), 20.66 MiB | 6.17 MiB/s, done.
Resolving deltas: 100% (23374/23374), done.

# mv etherpad-lite test-etherpad

# cd test-etherpad

# bin/run.sh
You shouldn't start Etherpad as root!
Please type 'Etherpad rocks my socks' or supply the '--root' argument if you still want to start it as root
Etherpad rocks my socks
bin/run.sh: 24: [: Etherpad rocks my socks: unexpected operator
Ensure that all dependencies are up to date...  If this is the first time you have run Etherpad please be patient.
npm WARN deprecated superagent@3.8.3: Please note that v5.0.1+ of superagent removes User-Agent header by default, therefore you may need to add it yourself (e.g. GitHub blocks requests without a User-Agent header).  This notice will go away with v5.0.2+ once it is released.
npm WARN deprecated npm-registry-fetch@1.1.1: this version has a breaking change. use 1.1.0 or upgrade to latest
npm WARN deprecated socks@1.1.10: If using 2.x branch, please upgrade to at least 2.1.6 to avoid a serious bug with socket data flow and an import issue introduced in 2.1.0

> wd@1.11.1 install /var/www/etherpad/test-etherpad/src/node_modules/wd
> node scripts/build-browser-scripts

added 1063 packages in 22.066s
Clearing minified cache...
Started Etherpad...
[2019-08-20 12:10:40.136] [DEBUG] console - Running on Node v9.11.2 (minimum required Node version: 6.9.0)
[2019-08-20 12:10:40.169] [INFO] console - All relative paths will be interpreted relative to the identified Etherpad base dir: /var/www/etherpad/test-etherpad
[2019-08-20 12:10:40.170] [DEBUG] AbsolutePaths - Relative path "settings.json" can be rewritten to "/var/www/etherpad/test-etherpad/settings.json"
[2019-08-20 12:10:40.170] [DEBUG] AbsolutePaths - Relative path "credentials.json" can be rewritten to "/var/www/etherpad/test-etherpad/credentials.json"
[2019-08-20 12:10:40.171] [INFO] console - Settings loaded from: /var/www/etherpad/test-etherpad/settings.json
[2019-08-20 12:10:40.171] [INFO] console - No credentials file found in /var/www/etherpad/test-etherpad/credentials.json. Ignoring.
[2019-08-20 12:10:40.177] [INFO] console - Using skin "no-skin" in dir: /var/www/etherpad/test-etherpad/src/static/skins/no-skin
[2019-08-20 12:10:40.177] [INFO] console - Session key file "/var/www/etherpad/test-etherpad/SESSIONKEY.txt" not found. Creating with random contents.
[2019-08-20 12:10:40.178] [WARN] console - DirtyDB is used. This is fine for testing but not recommended for production. File location: /var/www/etherpad/test-etherpad/var/dirty.db
[2019-08-20 12:10:41.336] [INFO] APIHandler - Api key file "/var/www/etherpad/test-etherpad/APIKEY.txt" not found. Creating with random contents.
[2019-08-20 12:10:41.650] [INFO] console - Installed plugins: 
[2019-08-20 12:10:41.654] [INFO] console - Report bugs at https://github.com/ether/etherpad-lite/issues
[2019-08-20 12:10:41.655] [INFO] console - Your Etherpad version is 1.7.5 (6a3e4c6)
[2019-08-20 12:10:41.778] [INFO] console - You can access your Etherpad instance at http://0.0.0.0:9003/
[2019-08-20 12:10:41.778] [INFO] console - The plugin admin page is at http://0.0.0.0:9003/admin/plugins
[2019-08-20 12:10:41.778] [WARN] console - Etherpad is running in Development mode.  This mode is slower for users and less secure than production mode.  You should set the NODE_ENV environment variable to production by using: export NODE_ENV=production
npm WARN saveError ENOENT: no such file or directory, open '/var/www/etherpad/test-etherpad/package.json'
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN enoent ENOENT: no such file or directory, open '/var/www/etherpad/test-etherpad/package.json'
npm WARN test-etherpad No description
npm WARN test-etherpad No repository field.
npm WARN test-etherpad No README data
npm WARN test-etherpad No license field.
[2019-08-20 12:12:10.170] [ERROR] console -

[2019-08-20 12:12:10.172] [INFO] console - + ep_small_list@0.3.0
added 1 package from 1 contributor and audited 5701 packages in 8.969s
[2019-08-20 12:12:10.173] [INFO] console - found 13 vulnerabilities (2 low, 11 high)
  run `npm audit fix` to fix them, or `npm audit` for details
[2019-08-20 12:12:10.986] [INFO] console - Restarting express server
[2019-08-20 12:12:43.982] [ERROR] console - TypeError: Cannot read property 'then' of undefined
    at /var/www/etherpad/test-etherpad/node_modules/ep_small_list/index.js:8:33
    at Layer.handle [as handle_request] (/var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/layer.js:95:5)
    at next (/var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/layer.js:95:5)
    at /var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/index.js:281:22
    at Function.process_params (/var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/index.js:335:12)
    at next (/var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/index.js:275:10)
    at /var/www/etherpad/test-etherpad/src/node/hooks/express/webaccess.js:85:20
    at authorize (/var/www/etherpad/test-etherpad/src/node/hooks/express/webaccess.js:24:51)
    at exports.basicAuth (/var/www/etherpad/test-etherpad/src/node/hooks/express/webaccess.js:84:3)
    at Layer.handle [as handle_request] (/var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/index.js:317:13)
    at /var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/index.js:335:12)
    at next (/var/www/etherpad/test-etherpad/src/node_modules/express/lib/router/index.js:275:10)
[            ......] | postinstall: sill install printInstalled`

**Comment by spruce** [comment]
Okay. Now I can reproduce and now what's going wrong but not quite understand why it works in my installation. Gonna try and create a mitigation for it.

**Comment by spruce** [comment]
Can you replace `./node_modules/ep_small_list/index.js` with the following content:
```
var padManager = require('ep_etherpad-lite/node/db/PadManager');
exports.eejsBlock_indexWrapper = function(hook_name, args, cb) {
    args.content += '<div id="small_list"></div><script src="./static/js/jquery.js"></script><script>$(function () { $("#small_list").load("./small_list"); });</script>';
};

exports.registerRoute = function(hook_name, args, cb) {
    args.app.get("/small_list", function(req, res) {
      var pads = padManager.listAllPads(createCallback(res));
      if(pads){
        pads.then(createCallback(res)).catch(function(err){
            r = "<div> Error:" + err + "</div>";
            res.send(r);
        })
      }
    });
};

function createList(data){
  r = "<ul>";
  if(data && data.padIDs){
    for (var i = 0; i < data.padIDs.length; i++) {
        r += '<li><a href="./p/' + data.padIDs[i] + '">' + data.padIDs[i] + '</li>';
    }
  }
  r += "</ul>";
  return r;
}

function createCallback(res){
  return function(err, data){
    if(err != null){
      data = err;
    }
    var val = createList(data);
    res.send(val);
  };
}
```

**Comment by mschmidt42** [comment]
Great. Now it works for me!

**Comment by spruce** [comment]
Should be fixed as of c6e2349fb367185f2cc68ff9ee1546879f2e8d79 which was released with 0.4.0 on npm.
